android/src/main/java/com/mattermost/networkclient/APIClientModule.kt

```diff
@@ -2,25 +2,52 @@ package com.mattermost.networkclient
 
 import com.mattermost.networkclient.enums.APIClientEvents
 import com.mattermost.networkclient.enums.RetryTypes
+import com.mattermost.networkclient.helpers.ProgressListener
+import com.mattermost.networkclient.helpers.UploadFileRequestBody
 import com.facebook.react.bridge.*
+import com.facebook.react.modules.network.ForwardingCookieHandler
+import com.facebook.react.modules.network.ReactCookieJarContainer
 import okhttp3.*
 import okhttp3.HttpUrl.Companion.toHttpUrl
+import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
 import android.net.Uri
-import android.util.Log
-import com.mattermost.networkclient.helpers.ProgressListener
-import com.mattermost.networkclient.helpers.UploadFileRequestBody
-import java.io.IOException
 import java.lang.Exception
 import kotlin.collections.HashMap
 
 class APIClientModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
-    private val clients = mutableMapOf<HttpUrl, NetworkClient>()
-    private val calls = mutableMapOf<String, Call>()
-
     override fun getName(): String {
         return "APIClient"
     }
 
+    companion object {
+        private val clients = mutableMapOf<HttpUrl, NetworkClient>()
+        private val calls = mutableMapOf<String, Call>()
+        val cookieJar = ReactCookieJarContainer()
+
+        fun getClientForRequest(request: Request): NetworkClient? {
+            var urlParts = request.url.toString().split("/")
+            while (urlParts.isNotEmpty()) {
+                val url = urlParts.joinToString (separator = "/") { it }.toHttpUrlOrNull()
+                if (url !== null && clients.containsKey(url)) {
+                    return clients[url]!!
+                }
+
+                urlParts = urlParts.dropLast(1)
+            }
+
+            return null
```

Comment:
Just wondering if it's worth starting from the first `urlPart` and adding to it rather than starting with the whole one and dropping parts? I'm guessing `baseUrl` will more often than not just be `domain.tld`